### PR TITLE
Use UUIDs instead of object hashes to avoid collisions

### PR DESCRIPTION
**Pull request by AnandInguva**
Fixes: https://github.com/apache/beam/issues/29600

Elements before CoGroupByKey with the same hash are not grouped together and treated as duplicates. Add unique id to the hash to prevent collisions. 

We should be creating unique id in the first place instead of hash object

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: sdks/python/apache_beam/ml/transforms/handlers.py [hunk 1]
```diff
@@ -134,7 +135,9 @@ def process(self, element):
         hash_object.update(str(list(value)).encode())
       else:  # assume value is a primitive that can be turned into str
         hash_object.update(str(value).encode())
-    yield (hash_object.hexdigest(), element)
+    # add a unique suffix to the hash key to avoid collisions.
+    unique_suffix = uuid.uuid4().hex
```
Comment (AnandInguva) [inline sdks/python/apache_beam/ml/transforms/handlers.py:139]: Thinking about this, we can just attach a unique id to each element. We just need unique id so that we can group transformed pcoll with untransformed pcoll with the downstream CoGroupByKey Pcoll

Comment (tvalentyn) [reply sdks/python/apache_beam/ml/transforms/handlers.py:139]: > Thinking about this, we can just attach a unique id to each element.

Not sure how accurate this is:
https://stackoverflow.com/questions/72989272/python-generates-same-uuid-over-multiple-docker-containers
  but I was also thinking about how the seed will be initialized in docker context.

and UUID1 might be not thread-safe:
https://github.com/ClickHouse/clickhouse-connect/issues/194

I'd expect the combination of   uuid1 + os PID + uuid4  would be exceedingly unlikely to collide in beam/dataflow context even with the above consideration. we can and should detect collisions though, we can fail the pipeline if that happens.

A few other notes:

1. Please add a known issue to Changes.MD and file a GH issue that will be fixed by this PR.
2. Let's rename the ComputeAndAttachHashKey to ComputeAndAttachUniqueID
3. I wonder if performance and pipeline cost would improve if we  can find a way to pass-through columns that do not need to be processed to tft, converting them to bytes if necessary.

Comment (tvalentyn) [reply sdks/python/apache_beam/ml/transforms/handlers.py:139]: also i wonder what is the overhead of TFTransform compared to plain usage of TFT, although that's a separate question

Comment (tvalentyn) [reply sdks/python/apache_beam/ml/transforms/handlers.py:139]: also if keys can be `bytes` instead of `str`, using raw bytes instead of hash would cut key size in ~half

Comment (AnandInguva) [reply sdks/python/apache_beam/ml/transforms/handlers.py:140]: I think we shouldn't raise an error as of now since the chance of collision is very low.
 
>> I wonder if performance and pipeline cost would improve if we can find a way to pass-through columns that do not need to be processed to tft, converting them to bytes if necessary, avoiding the shuffle step.

I can try this and test the performance. If this works, we can remove the CoGroupByKey altogether.

File: sdks/python/apache_beam/ml/transforms/handlers.py [hunk 2]
```diff
@@ -134,7 +137,11 @@ def process(self, element):
         hash_object.update(str(list(value)).encode())
       else:  # assume value is a primitive that can be turned into str
         hash_object.update(str(value).encode())
-    yield (hash_object.hexdigest(), element)
+    unique_suffix = uuid.uuid4().hex
+    # over multpile docker containers, the uuid might generate
+    # same uuid. So we will add the process id to the hash key
+    # along with the unique suffix to avoid possible collisions.
+    yield (hash_object.digest() + str(os.getpid()) + unique_suffix, element)
```
[the remark this answers is not in the document]
Comment (AnandInguva) [reply sdks/python/apache_beam/ml/transforms/handlers.py]: Changed it.

**Comment by AnandInguva** [comment]
I will create a github issue to track >> I wonder if performance and pipeline cost would improve if we can find a way to pass-through columns that do not need to be processed to tft, converting them to bytes if necessary, avoiding the shuffle step. 

Performance testing MLTransform should help here.

File: sdks/python/apache_beam/ml/transforms/handlers.py [hunk 3]
```diff
@@ -201,6 +198,8 @@ class MergeDicts(beam.DoFn):
   def process(self, element):
     _, element = element
     new_dict = {}
+    # Assertion could fail due to UUID collision.
```
Comment (tvalentyn) [inline sdks/python/apache_beam/ml/transforms/handlers.py:201]: @AnandInguva please check that this condition makes sense. In particular, are there exactly two elements per key or two elements per each transformation?

**Comment by tvalentyn** [comment]
added some suggestions, ptal and run MLTransform integration tests. Thanks.

File: sdks/python/apache_beam/ml/transforms/handlers.py (hunk 3 above)
Comment (AnandInguva) [reply sdks/python/apache_beam/ml/transforms/handlers.py:201]: It would be per key. The elements len might be greater than 2. Changing the condition a bit.

File: sdks/python/apache_beam/examples/snippets/transforms/elementwise/mltransform_test.py [hunk 4]
```diff
@@ -57,6 +57,7 @@ def check_mltransform_compute_and_apply_vocabulary_with_scalar():
 Row(x=array([4]))
 Row(x=array([1]))
 Row(x=array([0]))
+Row(x=array([0]))
```
Comment (tvalentyn) [inline sdks/python/apache_beam/examples/snippets/transforms/elementwise/mltransform_test.py:60]: we also need to replicate this on the website, can be a separate PR.

Comment (AnandInguva) [reply sdks/python/apache_beam/examples/snippets/transforms/elementwise/mltransform_test.py:60]: this is the same example that will be updated on the website with this change.

Comment (tvalentyn) [reply sdks/python/apache_beam/examples/snippets/transforms/elementwise/mltransform_test.py:60]: ah, yes this is a snippet. thanks

File: sdks/python/apache_beam/ml/transforms/handlers.py (hunk 3 above)
Comment (tvalentyn) [reply sdks/python/apache_beam/ml/transforms/handlers.py]: Thanks, I made a few clarifying edits, PTAL.

**Review by tvalentyn — APPROVED** [review]

**Comment by tvalentyn** [comment]
@AnandInguva please merge if tests pass and new commits look good to you.

File: sdks/python/apache_beam/ml/transforms/handlers.py [hunk 5]
```diff
@@ -121,76 +123,74 @@ def expand(
...
   def process(self, element):
-    new_dict = {}
-    hash_key, element = element
-    for key, value in element.items():
-      if key not in self.existing_columns:
-        new_dict[key] = value
-    yield (hash_key, new_dict)
+    id, row_dict = element
+    new_dict = {
+        k:v for k, v in row_dict.items() if k not in self.existing_columns
```
Comment (tvalentyn) [inline sdks/python/apache_beam/ml/transforms/handlers.py:157]:
```suggestion
        k: v
        for k, v in row_dict.items() if k not in self.existing_columns
```

**Comment by AnandInguva** [comment]
lgtm
